### ISSUE TITLE
Calculation of "keep" for harvest missions

### ISSUE THREAD

**Issue by thraidh** [Mmtrx/FS22_BetterContracts#4]
Since patch 1.3 the factor of actually expected to the announced seems to have changed from 0.735 to 0.9069. Therefore the currently displayed profits are way too high.

**Comment by Mmtrx** [comment]
Hi, I just checked the success factors for harvest missions in the std game. They did not change with 1.3.0. 
To get the max "keep", you need to completely harvest the field first, before you start delivering to the sellpoint. 
Or are you talking about mowing / baling missions? Same there: first get all grass mowed, before starting to deliver bales.

**Comment by thraidh** [comment]
Well... that would be hard for e.g. potatoes or big barley fields, which were the things I tried. Anyway, before the patch everything was fine. After the patch, if you deliver everything, so the mission is completed, do not hand it in, but save and check missions.xml. For two different missions and crops, the percentage between delivered and expected was around 90.68something% both times.

**Comment by Mmtrx** [comment]
Can you paste the part of missions.xml here? Maybe with a screen shot of the contracts screen for the active mission, details on, so that we can see the field completion value?

**Comment by thraidh** [comment]
I'm on Haut-Beyleron, field 33, canola:
```xml
    <mission type="harvest" activeId="3" reward="6295" status="1" success="false" farmId="1">
        <field id="33" sprayFactor="0.000000" spraySet="false" plowFactor="1.000000" state="2" vehicleGroup="16" vehicleUseCost="839.340027" spawnedVehicles="false" growthState="9" limeFactor="0.000000" weedFactor="1.000000" stubbleFactor="0.000000" fruitTypeName="CANOLA"/>
        <harvest sellPointPlaceableId="21" unloadingStationIndex="1" expectedLiters="32860.160156" depositedLiters="0.000000"/>
    </mission>
```

This is after the field is fully harvested (33229l), but nothing deliverd. The expected amount is displayed as 24200l, price is 1729 Euro, expected profit 15059 Euro. The completion value is 80% (this was 84% before the patch).

This is after everything was delivered with an extra profit of 5658 Euro:
```xml
    <mission type="harvest" activeId="3" reward="6295" status="2" success="true" farmId="1" stealingCost="0.000000">
        <field id="33" sprayFactor="0.000000" spraySet="false" plowFactor="1.000000" state="2" vehicleGroup="16" vehicleUseCost="839.340027" spawnedVehicles="false" growthState="9" limeFactor="0.000000" weedFactor="1.000000" stubbleFactor="0.000000" fruitTypeName="CANOLA"/>
        <harvest sellPointPlaceableId="21" unloadingStationIndex="1" expectedLiters="32860.160156" depositedLiters="29959.177734"/>
    </mission>
```

Here you can also see: `depositedLiters / expectedLiters ~ 0.91`, which is even worse than the 0.907 I had before.

**Comment by thraidh** [comment]
Here is another completed one:
```xml
    <mission type="harvest" activeId="5" reward="1605" status="2" success="true" farmId="1" stealingCost="0.000000">
        <field id="23" sprayFactor="1.000000" spraySet="false" plowFactor="1.000000" state="2" vehicleGroup="12" vehicleUseCost="214.119995" spawnedVehicles="false" growthState="5" limeFactor="1.000000" weedFactor="1.000000" stubbleFactor="0.000000" fruitTypeName="SORGHUM"/>
        <harvest sellPointPlaceableId="12" unloadingStationIndex="1" expectedLiters="17118.894531" depositedLiters="15589.631836"/>
    </mission>
```

**Comment by thraidh** [comment]
And one more:
```xml
    <mission type="harvest" activeId="3" reward="10343" status="1" success="false" farmId="1">
        <field id="36" sprayFactor="1.000000" spraySet="false" plowFactor="1.000000" state="2" vehicleGroup="3" vehicleUseCost="1379.180054" spawnedVehicles="false" growthState="5" limeFactor="1.000000" weedFactor="1.000000" stubbleFactor="0.000000" fruitTypeName="SORGHUM"/>
        <harvest sellPointPlaceableId="12" unloadingStationIndex="1" expectedLiters="111679.101562" depositedLiters="106099.640625"/>
    </mission>
```
This is even worse at 95% expected.

**Comment by Mmtrx** [comment]
Ok, you are right. They didn't changed the success multipliers, but slightly the way of computing when mission is complete. Formula is `0.8 * harvestCompletion + 0.2 * sellCompletion`, where `harvestCompletion = fieldcompletion / 0.95` So you could "overachieve" the harvest part, and get a harvestCompletion of 1/.095 = 1.05. Leading to a mission completion of 0.8 * 1.05 = 0.84. Then you only had to reach a sellCompletion of 80%.

In 1.3.0. they cap the harvestCompletion at 100%, so you need to really reach 100% of sellCompletion.
`sellCompletion = depositedLiters / expectedLiters / 0.93  --> depositedLiters / expectedLiters = 0.93`  

I will change the calculation in the mod with next release.

**Comment by thraidh** [comment]
Actually it seems that `sellCompletion=depositedLiters / (expectedLiters * 0.95)`. At least that is what I observed, when I compared the percentages displayed against the numbers from `missions.xml`.
That would also mean that you'd get more or less exactly 95%, if you sell before you harvest, but if you do that, `fieldcompletion` is less than if you harvest first.
I guess the total is capped, but you can actually overachieve in harvesting or selling, thereby reducing the amount you need of the other one.
I'm not exactly sure about the formula, though. Did you get the above numbers from some source or are those also just observations?

Here is another one, where I first harvested the whole potato field and then sold everything:
```xml
    <mission type="harvest" activeId="1" reward="18096" status="2" success="true" farmId="1" stealingCost="0.000000">
        <field id="45" sprayFactor="1.000000" spraySet="false" plowFactor="1.000000" state="2" vehicleGroup="12" vehicleUseCost="754.000000" spawnedVehicles="true" growthState="6" limeFactor="1.000000" weedFactor="1.000000" stubbleFactor="0.000000" fruitTypeName="POTATO"/>
        <harvest sellPointPlaceableId="19" unloadingStationIndex="1" expectedLiters="307509.468750" depositedLiters="278953.093750"/>
    </mission>
```
At 20800l delivered it showed 71% delivered, at 260000l it showed 89%, which would mean that 100% is around 292000l, which would be 95% of `expectedLiters`. But the whole contract was fulfilled at 278953l (which would be 90.7%).

**Comment by Mmtrx** [comment]
What the game displays as contract completion (the percentage bar) is always the combined completion rate, `0.8 * harvestCompletion + 0.2 * sellCompletion`. So you cannot directly compare the deposited / expected from missions.xml with the displayed overall contract completion.

And no, since 1.3.0 you no longer can overachieve either of the two. The game calculates completion of a harvest contract as:
```
harvestCompletion = min (fieldcompletion / 0.95, 1)
sellCompletion = min (depositedLiters / expectedLiters / 0.93, 1)
contractCompletion = min (0.8 * harvestCompletion + 0.2 * sellCompletion, 1)
``` 
Because of the min(), both harvest and sell can never get over 100%. (In 1.2.0, the min() was not there.)
(Yep, the formulas are directly from the game lua :)

**Comment by thraidh** [comment]
Mystery solved: apparently the contract is complete as soon as `contractCompletion` is rounded to 100%, so `>= 0.995`. That can actually make a difference of a few thousand liters and change the required percentage by a few points.

**Comment by Plopsi** [comment]
I just tested on Nordfriesische Marsch - the value to completion (after 100% field completion) is 90,6751% (90,6750004651425% to be exact +/- 0,0000000000001%) - tested from 1000l, 10kl, 100kl and 123.456.789 liters.

- Plopsi

**Comment by Mmtrx** [comment]
Calculation fixed, according to FS22 v1.3.1.0. See [Release](https://github.com/Mmtrx/FS22_BetterContracts/releases/tag/v1.2.1.0)
